harness/determined/pytorch/_pytorch_context.py

```diff
@@ -57,4 +102,359 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+            else:
+                raise det.errors.InvalidExperimentException(
+                    "Calling context.backward(loss) multiple times is not supported "
+                    "while using apex.amp and parallel/distributed training"
+                )
+
+            if loss not in self._loss_ids:
+                self._loss_ids[loss] = len(self._loss_ids)
+            with apex.amp.scale_loss(
+                loss, self.optimizers, loss_id=self._loss_ids[loss]
+            ) as scaled_loss:
+                scaled_loss.backward(
+                    gradient=gradient, retain_graph=retain_graph, create_graph=create_graph
+                )
+
+                if self.hvd_config.use and self._should_communicate_and_update():
+                    # When using horovod, we need to finish communicating gradient
+                    # updates before they are unscaled which happens when we exit
+                    # of this context manager.
+                    #
+                    # Unfortunately, there is no clean way to support unscaling
+                    # happening after synchronizing the optimizer on apex.amp.
+                    # A short-term solution is to not support multiple losses nor
+                    # multiple backward passes on one loss. A long-term solution is
+                    # to integrate torch native AMP (https://pytorch.org/docs/stable/amp.html),
+                    # which will come out soon.
+                    for optimizer in self.optimizers:
+                        optimizer.synchronize()
+        else:
+            loss.backward(  # type: ignore
+                gradient=gradient, retain_graph=retain_graph, create_graph=create_graph,
+            )
+
+    def _average_gradients(self, parameters: Any, divisor: int) -> None:
+        check.gt_eq(divisor, 1)
+        if divisor == 1:
+            return
+
+        divisor_value = float(divisor)
+        for p in filter(lambda param: param.grad is not None, parameters):
+            p.grad.data.div_(divisor_value)
+
+    def _step_optimizer(
+        self,
+        optimizer: torch.optim.Optimizer,  # type: ignore
+        clip_grads: Optional[Callable[[Iterator], None]] = None,
+    ) -> None:
+        """
+        Perform a single optimization step.
+
+        This function must be called once for each optimizer. However, the order of
+        different optimizers' optimization steps can be specified by calling
+        this function in different orders. Also, gradient accumulation across iterations
+        is performed by the Determined training loop by setting the experiment configuration
+        optimization.aggregation_frequency.
+
+        Arguments:
+            optimizer(``torch.optim.Optimizer``): Which optimizer should be stepped.
+            clip_grads(a function, optional): This function should have one argument for
+                parameters in order to clip the gradients
+        """
+
```

Comment:
question: is there any mechanism for checking if users have called backwards for all their optimizers once per step? And what happens if they skip an optimizer?

Reply:
There is no mechanism for that. We don't support skipping optimizer.step. That is equivalent to manual gradient accumulation. Will add checks for that.

Reply:
Nice catch!

Reply:
But the horovod will throw the same errors when users skip calling `optimizer.step` as when users do manual gradient accumulation.